draft-irtf-cfrg-hybrid-kems.md

```diff
@@ -537,175 +545,326 @@ and L arguments.
+
+## Subroutines
+
+The four hybrid KEM framework share a substantial amount of structure, which we
+capture in a set of subroutines.
 
-Given these constituent parts, the GHP hybrid KEM is defined as follows:
+### Using a Nominal Group
+
+Hybrid KEM frameworks that use a KEM for the traditional component invoke the
+DeriveKeyPair, Encap, and Decap functions of PQ KEMs, alongside analogous
+functions of the nominal group.  The "encapsulation key" is the receiver's
+public group element; the "ciphertext" is an ephemeral group element; and
+the shared secret is the secret value resulting from an ephemeral-static
+Diffie-Hellman exchange.
+
+~~~ pseudocode
+def expandDecapsKeyG(seed):
+    seed_full = PRG(seed)
+    (seed_PQ, seed_T) = split(KEM_PQ.Nseed, Group_T.Nseed, seed_full)
+
+    (ek_PQ, dk_PQ) = KEM_PQ.DeriveKeyPair(seed_PQ)
+    dk_T = Group_T.RandomScalar(seed_T)
+    ek_T = Group_T.Exp(Group_T.g, dk_T)
 
+    return (ek_PQ, ek_T, dk_PQ, dk_T)
+
+def prepareEncapG(ek_PQ, ek_T):
+    (ss_PQ, ct_PQ) = KEM_PQ.Encap(ek_PQ)
+    sk_E = Group_T.RandomScalar(random(Group_T.Nseed))
+    ct_T = Group_T.Exp(Group_T.g, sk_E)
+    ss_T = Group_T.ElementToSharedSecret(Group_T.Exp(ek_T, sk_E))
+    return (ss_PQ, ss_T, ct_PQ, ct_T)
+
+def prepareDecapG(ct_PQ, ct_T, dk_PQ, dk_T):
+    ss_PQ = KEM_PQ.Decap(dk_PQ, ct_PQ)
+    ss_T = Group_T.ElementToSharedSecret(Group_T.Exp(ct_T, dk_T))
+    return (ss_PQ, ss_T)
 ~~~
-def expandDecapsulationKey(seed):
+
+### Using a Traditional KEM
+
+Hybrid KEM frameworks that use a KEM for the traditional component invoke the
+DeriveKeyPair, Encap, and Decap functions of the traditional and PQ KEMs in
+parallel.
+
+~~~ pseudocode
+def expandDecapsKeyK(seed):
     seed_full = PRG(seed)
     (seed_PQ, seed_T) = split(KEM_PQ.Nseed, KEM_T.Nseed, seed_full)
     (ek_PQ, dk_PQ) = KEM_PQ.DeriveKeyPair(seed_PQ)
     (ek_T, dk_T) = KEM_T.DeriveKeyPair(seed_T)
     return (ek_PQ, ek_T, dk_PQ, dk_T)
 
-def DeriveKeyPair(seed):
-    (ek_PQ, ek_T, dk_PQ, dk_T) = expandDecapsulationKey(seed)
-    return (concat(ek_PQ, ek_T), seed)
-
-def GenerateKeyPair():
-    seed = random(Nseed)
-    return DeriveKeyPair(seed)
-
-def Encaps(ek):
-    (ek_PQ, ek_T) = split(KEM_PQ.Nek, KEM_T.Nek, ek)
+def prepareEncapK(ek_PQ, ek_T):
     (ss_PQ, ct_PQ) = KEM_PQ.Encap(ek_PQ)
     (ss_T, ct_T) = KEM_T.Encap(ek_T)
-    ss_H = KDF(concat(ss_PQ, ss_T, ct_PQ, ct_T, ek_PQ, ek_T, label))
-    ct_H = concat(ct_PQ, ct_T)
-    return (ss_H, ct_H)
-
-def Decaps(dk, ct):
-    (ek_PQ, ek_T, dk_PQ, dk_T) = expandDecapsulationKey(dk)
+    return (ss_PQ, ss_T, ct_PQ, ct_T)
 
-    (ct_PQ, ct_T) = split(KEM_PQ.Nct, KEM_T.Nct, ct)
+def prepareDecapK(ct_PQ, ct_T, dk_PQ, dk_T):
     ss_PQ = KEM_PQ.Decap(dk_PQ, ct_PQ)
     ss_T = KEM_T.Decap(dk_T, ct_T)
+    return (ss_PQ, ss_T)
+~~~
+
+### Combiners
+
+A combiner function uses the KDF used in the hybrid KEM to combine the shared
+secrets output by the constituent algorithms with metadata.  This operation is
```

Comment:
``` suggestion
secrets output by the component algorithms with contextual information. This operation is
```

You say "component", I say "constituent", let's pick one and not both :)  Seems like "component is the one more people like.